hudi-client/src/main/java/org/apache/hudi/table/HoodieCopyOnWriteTable.java

```diff
@@ -602,18 +602,39 @@ private int addUpdateBucket(String fileIdHint) {
       return bucket;
     }
 
-    private void assignInserts(WorkloadProfile profile) {
+    private void assignInserts(WorkloadProfile profile, JavaSparkContext jsc) {
       // for new inserts, compute buckets depending on how many records we have for each partition
       Set<String> partitionPaths = profile.getPartitionPaths();
       long averageRecordSize =
           averageBytesPerRecord(metaClient.getActiveTimeline().getCommitTimeline().filterCompletedInstants(),
               config.getCopyOnWriteRecordSizeEstimate());
       LOG.info("AvgRecordSize => " + averageRecordSize);
+
+      HashMap<String, List<SmallFile>> partitionSmallFilesMap = new HashMap<>();
+      if (jsc != null && partitionPaths.size() > 1) {
+        //Parellelize the GetSmallFile Operation by using RDDs
+        List<String> partitionPathsList = new ArrayList<>(partitionPaths);
+        JavaRDD<String> partitionPathRdds = jsc.parallelize(partitionPathsList, partitionPathsList.size());
+        List<Tuple2<String, List<SmallFile>>> partitionSmallFileTuples =
+                partitionPathRdds.map(it -> new Tuple2<String, List<SmallFile>>(it, getSmallFiles(it))).collect();
+
+        for (Tuple2<String, List<SmallFile>> tuple : partitionSmallFileTuples) {
+          partitionSmallFilesMap.put(tuple._1, tuple._2);
+        }
+      }
+
       for (String partitionPath : partitionPaths) {
         WorkloadStat pStat = profile.getWorkloadStat(partitionPath);
         if (pStat.getNumInserts() > 0) {
 
-          List<SmallFile> smallFiles = getSmallFiles(partitionPath);
+          List<SmallFile> smallFiles;
+          if (partitionSmallFilesMap.isEmpty()) {
```

Comment:
Can we change that code so that it always lists in parallel.. i.e no need for the fallback in this if block...

Reply:
sounds good. will do